src/tools/clippy/clippy_utils/src/lib.rs

```diff
@@ -314,7 +314,13 @@ pub fn is_diag_item_method(cx: &LateContext<'_>, def_id: DefId, diag_item: Symbo
 /// Checks if a method is in a diagnostic item trait
 pub fn is_diag_trait_item(cx: &LateContext<'_>, def_id: DefId, diag_item: Symbol) -> bool {
     if let Some(trait_did) = cx.tcx.trait_of_item(def_id) {
-        return cx.tcx.is_diagnostic_item(diag_item, trait_did);
+        // `Iterator` is a lang item, not a diag item, but we'd like to check for `Iterator` methods
+        // the same way we check for other traits' methods.
+        if diag_item == sym::Iterator {
+            return cx.tcx.lang_items().iterator_trait() == Some(trait_did)
+        } else {
+            return cx.tcx.is_diagnostic_item(diag_item, trait_did);
+        }
```

Comment:
WDYT about making `Iterator` a diag item and a lang item at the same time. This change feels a bit hacky to me - singling out the `Iterator` trait.

Reply:
https://github.com/rust-lang/rust/pull/116447#discussion_r1368533339

Reply:
done